### PR TITLE
[WIP] metastack

### PR DESCRIPTION
**Pull request by trehn**

### CODE REVIEW

File: bundlewrap/utils/metastack.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+        for layer in [self._base] + list(self._layers.values()):
+            exists, value = self._dict_has_path(layer, path)
+            if exists:
+                if undef:
+                    # First time we see anything.
+                    result = {'data': value}
+                    undef = False
+                else:
+                    result = merge_dict(result, {'data': value})
```
Comment (vain) [inline bundlewrap/utils/metastack.py]: check for type conflicts

File: bundlewrap/utils/metastack.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        if undef:
+            if use_default:
+                return default
+            else:
+                raise MetastackKeyError(f'Path {path} not in metastack')
+        else:
+            return result['data']
```
Comment (vain) [inline bundlewrap/utils/metastack.py]: freeze everything we return from this method

File: bundlewrap/utils/metastack.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            current = current[element]
+
+        return True, current
+
+    def _merge_layers(self, base, update):
+        merged = base.copy()
+
+        for key, value in update.items():
+            if isinstance(value, _Atomic):
+                raise MetastackAtomicIllegal()
```
Comment (vain) [inline bundlewrap/utils/metastack.py]: need to medidate and talk about this

File: bundlewrap/utils/dicts.py [hunk 4]
```diff
@@ -183,6 +191,29 @@ def default(self, obj):
...
+        for i in obj:
+            result.append(freeze_object(i))
+        return tuple(result)
+    elif isinstance(obj, set):
+        result = set()
+        for i in obj:
+            result.add(freeze_object(i))
+        return frozenset(obj)
+    else:
+        return obj
```
Comment (trehn) [inline bundlewrap/utils/dicts.py]: we could use this opportunity to check for other invalid types (see https://docs.bundlewrap.org/repo/nodes.py/#metadata)

**Pull request merged**